Libraries/LibMedia/Audio/PlaybackStreamWasapi.cpp

```diff
@@ -0,0 +1,502 @@
+}
+
+PlaybackStreamWasapi::PlaybackStreamWasapi(NonnullRefPtr<AudioState> state)
+    : m_state(move(state))
+{
+}
+
+PlaybackStreamWasapi::~PlaybackStreamWasapi()
+{
+    m_state->exit_requested.store(true, MemoryOrder::memory_order_release);
+    // Poke the event to wake the thread up from wait
+    if (m_state->buffer_event)
+        SetEvent(m_state->buffer_event);
+
+    MUST(m_state->audio_thread->join());
+}
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStream::create(OutputState initial_output_state, u32 sample_rate, u8 channels, u32 target_latency_ms, AudioDataRequestCallback&& callback)
+{
+    return PlaybackStreamWasapi::create(initial_output_state, sample_rate, channels, target_latency_ms, move(callback));
+}
+
+static void print_audio_format(WAVEFORMATEXTENSIBLE* format)
+{
+    dbgln("wFormatTag: {:x}\n"
+          "nChannels: {}\n"
+          "nSamplesPerSec: {}\n"
+          "nAvgBytesPerSec: {}\n"
+          "nBlockAlign: {}\n"
+          "wBitsPerSample: {}\n"
+          "cbSize: {}\n"
+          "Samples.wValidBitsPerSample: {}\n"
+          "dwChannelMask: {:b}\n"
+          "SubFormat: {}",
+        format->Format.wFormatTag,
+        format->Format.nChannels,
+        format->Format.nSamplesPerSec,
+        format->Format.nAvgBytesPerSec,
+        format->Format.nBlockAlign,
+        format->Format.wBitsPerSample,
+        format->Format.cbSize,
+        format->Samples.wValidBitsPerSample,
+        format->dwChannelMask,
+        Span<u8> { reinterpret_cast<char*>(&format->SubFormat), 16 });
+}
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamWasapi::create(OutputState initial_output_state, u32 sample_rate, u8 channels, u32, AudioDataRequestCallback&& data_request_callback)
+{
+    HRESULT hr;
+    if (!com_helper.initialized) {
+        hr = CoInitializeEx(NULL, COINIT_MULTITHREADED);
+        if (FAILED(hr))
+            return Error::from_windows_error(hr);
+        com_helper.initialized = true;
+    }
+
+    auto state = make_ref_counted<PlaybackStreamWasapi::AudioState>();
+
+    hr = CoCreateInstance(__uuidof(MMDeviceEnumerator), NULL, CLSCTX_ALL, IID_PPV_ARGS(&state->enumerator));
+    if (FAILED(hr))
+        return Error::from_windows_error(hr);
+
+    hr = state->enumerator->GetDefaultAudioEndpoint(eRender, eConsole, &state->device);
+    if (FAILED(hr))
+        return Error::from_windows_error(hr);
+
+    hr = state->device->Activate(__uuidof(IAudioClient), CLSCTX_ALL, nullptr, &state->audio_client);
+    if (FAILED(hr))
+        return Error::from_windows_error(hr);
+
+    state->data_request_callback = move(data_request_callback);
+
+    WAVEFORMATEXTENSIBLE* ideal_format;
+    state->audio_client->GetMixFormat(reinterpret_cast<WAVEFORMATEX**>(&ideal_format));
+    dbgln("Mixing engine audio format:\n");
+    print_audio_format(ideal_format);
+    CoTaskMemFree(ideal_format);
+    // Set up 32bit float PCM
+    u16 bits_per_sample = 32;
+    u16 block_align = (channels * bits_per_sample) / 8;
```

Comment:
Would be nice to make this a bit clearer:

```suggestion
    u16 bits_per_sample = sizeof(float) * __CHAR_BIT__;
    u16 block_align = channels * sizeof(float);
```